### PR TITLE
Fix: WASM Migration Phase 8: Implement safeZoneAdjustment() in Rust/WASM for all activation functions (#1140)

### LINKED ISSUE #1140: WASM Migration Phase 8: Implement safeZoneAdjustment() in Rust/WASM for all activation functions
## Parent Issue
Part of #1136 - Replace all JS squash functions with Rust/WASM

## Background
Following Phase 7 (unSquash), this phase implements `safeZoneAdjustment()` in WASM. This function determines how suitable it is to backpropagate through a neuron based on saturation levels. It's critical for training stability.

## Scope
Implement `safeZoneAdjustment()` in Rust/WASM for all activation functions.

### Understanding safeZoneAdjustment
The function returns a float from 0 (not safe) to 1 (fully safe) indicating how useful it is to backpropagate through a neuron:
- **1.0**: Fully in safe zone, gradient flows freely
- **0.0**: Completely saturated, no gradient should flow
- **0.0-1.0**: Partial safety, used for gradual fade-out

### Functions to implement
Each activation has different safe zones based on their saturation behaviour:

| Function | Safe Zone (Full) | Fade Zone | Notes |
|----------|------------------|-----------|-------|
| Identity | (-∞, ∞) = 1.0 | None | Never saturates |
| ReLU | [0, ∞) = 1.0 | x < 0: recovery | Dead ReLU recovery |
| ReLU6 | (0, 6) = 1.0 | Edges | Both ends saturate |
| LeakyReLU | (-∞, ∞) = 1.0 | None | Never fully saturates |
| SELU/ELU | [-2, ∞) = 1.0 | x < -2: fade | Negative saturation |
| Logistic | [-6, 6] = full | ±6 to ±10 fade | Classic sigmoid |
| Tanh | [-2, 2] = full | ±2 to ±6 fade | Similar to logistic |
| HardTanh | (-1, 1) = 1.0 | At ±1 = 0 | Hard boundaries |
| Softsign | [-5, 5] = full | Fade to ±10 | Slow saturation |
| Softplus | [-10, ∞) = full | x < -10 fade | One-sided |
| Swish/Mish | Similar to tanh | | |
| GELU | Similar to ReLU | | Smoother transition |
| Sine/Cosine | Always varying | 1.0 | Periodic |
| Tan | Avoid ±π/2 | | Asymptotes |
| ArcTan | (-∞, ∞) | Fade at extremes | |
| Gaussian | [-3, 3] = full | Fade beyond | Bell curve |
| BentIdentity | (-∞, ∞) = 1.0 | None | Never saturates |
| BipolarSigmoid | [-6, 6] = full | Fade | |
| Bipolar/Step | 0.0 always | None | Discontinuous |
| Others | Per function | | |

### Implementation approach
1. Add `safe_zone_adjustment(squash_type: SquashType, raw_input: f32, error: f32) -> f32` in lib.rs
2. Implement recovery logic for each function (when error direction helps recovery)
3. Export via wasm_bindgen
4. Add TypeScript wrapper `wasmSafeZoneAdjustment(squashType: SquashType, rawInput: number, error: number): number`
5. Add tests comparing WASM and JS implementations

### Key patterns from JS implementations
```rust
// Pattern for bounded functions (like Tanh):
if raw_input >= safe_low && raw_input <= safe_high {
    return 1.0;  // Fully safe
}
// Recovery direction
if raw_input < safe_low && error > 0.0 {
    return 0.2;  // Allow small recovery gradient
}
// Gradual fade
if raw_input > safe_high && raw_input <= max {
    return 1.0 - (raw_input - safe_high) / (max - safe_high);
}
return 0.0;  // Fully saturated
```

### Acceptance Criteria
- [ ] All activation functions have safeZoneAdjustment implemented in Rust
- [ ] Recovery logic implemented for each function
- [ ] Exported via wasm_bindgen
- [ ] TypeScript wrapper function added
- [ ] Tests verify WASM and JS results match
- [ ] All existing tests pass
- [ ] `./quality.sh` passes

## Notes
- The recovery logic (allowing small gradients when error direction would help unsaturate) is critical for training
- Non-finite inputs should return 0
- Discontinuous functions (Bipolar, Step) should always return 0

### PR DESCRIPTION
**Pull request by stservice**
## Summary

This PR implements Phase 8 of the WASM Migration: `safeZoneAdjustment()` in
Rust/WASM for all activation functions.

The `safeZoneAdjustment()` function returns a float from 0 (not safe) to 1
(fully safe) indicating how suitable it is to backpropagate through a neuron
based on saturation levels. This is critical for training stability as it
prevents pushing already-saturated neurons further into saturation.

### Changes Made

1. **Rust Implementation** (`wasm_activation/src/lib.rs`):
   - Added `apply_safe_zone_adjustment()` function implementing safe zone logic
     for all 35 activation functions
   - Each activation function has customised logic based on its saturation
     behaviour:
     - **Identity, LeakyReLU, BentIdentity**: Nearly always safe, with
       weight-based adjustments
     - **ReLU, ReLU6**: Recovery logic for dead neurons
     - **TANH, Logistic, HardTanh**: Safe zones with gradual fade and recovery
     - **SELU, ELU, Swish, Mish, GELU, etc.**: Weight-based safety checks with
       fade zones
     - **SINE, Cosine, TAN**: Slope-based safety (periodic functions)
     - **BIPOLAR, Step**: Discontinuous functions with special handling
     - **Aggregate functions (Minimum, Maximum, If)**: Always return 0 (not
       differentiable)
   - Exported via wasm_bindgen as `safe_zone_adjustment()`

2. **TypeScript Wrapper** (`src/wasm/WasmActivation.ts`):
   - Added `wasmSafeZoneAdjustment()` function
   - Updated both `initWasmActivation()` and `initWasmActivationSync()` to load
     the new function
   - Added proper JSDoc documentation

3. **Module Export** (`src/wasm/mod.ts`):
   - Exported `wasmSafeZoneAdjustment` for external use

4. **Comprehensive Tests** (`test/WasmSafeZoneAdjustment.ts`):
   - 36 test cases covering all activation functions
   - Tests verify WASM results match JS implementations
   - Tests cover safe zones, recovery zones, saturation zones
   - Tests verify non-finite inputs return 0
   - Comprehensive comparison test: 1200/1200 passed

## Evidence

Unable to generate screenshot: This is a WASM/Rust implementation with no visual
interface. The implementation is verified through comprehensive unit tests.

### Test Results

All 36 tests pass, including:

- Individual tests for each of the 32 activation functions with
  safeZoneAdjustment
- Aggregate function tests (Minimum, Maximum, If return 0)
- Non-finite input handling tests
- Comprehensive comparison test (1200 combinations tested)

```
running 36 tests from ./test/WasmSafeZoneAdjustment.ts
WASM SafeZoneAdjustment: Module initialisation ... ok
WASM SafeZoneAdjustment: IDENTITY ... ok
WASM SafeZoneAdjustment: ReLU ... ok
[... all 36 tests pass ...]
Comprehensive SafeZoneAdjustment test: 1200/1200 passed

ok | 36 passed | 0 failed
```

Full quality.sh passes: `ok | 1673 passed (2 steps) | 0 failed | 1 ignored`

## Test Plan

- [x] Added `test/WasmSafeZoneAdjustment.ts` with comprehensive tests
- [x] Tests verify WASM and JS implementations match for all activation
      functions
- [x] Tests cover edge cases (non-finite inputs, boundary values, recovery
      conditions)
- [x] Tests verify aggregate functions return 0
- [x] Tests verify COMPLEMENT returns 1.0 (never saturates)
- [x] All existing tests continue to pass (1673 tests total)
- [x] `./quality.sh` passes

## Acceptance Criteria Checklist

- [x] All activation functions have safeZoneAdjustment implemented in Rust
- [x] Recovery logic implemented for each function
- [x] Exported via wasm_bindgen
- [x] TypeScript wrapper function added
- [x] Tests verify WASM and JS results match
- [x] All existing tests pass
- [x] `./quality.sh` passes

Fixes #1140

---

## Automated Fix Details
This PR addresses issue #1140: **WASM Migration Phase 8: Implement safeZoneAdjustment() in Rust/WASM for all activation functions**

## Quality Checks
- Followed TDD methodology
- All new functionality has corresponding tests
- Ran `./quality.sh` - all checks pass

## Notes
- No existing tests were removed or commented out
- If any test modifications were required due to business logic changes, they are documented in the commits

Closes #1140

### CODE REVIEW

**Comment by stservice** [comment]
Thanks for the automated dependency review!

No action required - this is a positive status report confirming:
- ✅ No vulnerabilities found
- ✅ No license issues found
- ✅ No OpenSSF Scorecard issues found

The PR is ready for review from a dependency security perspective.

**Comment by stservice** [comment]
Thanks for the automated test results report! 🎉

This is a positive status update confirming:
- ✅ **1,665 tests** executed
- ✅ **1,633 passed**
- 💤 **32 skipped** (expected behaviour for conditional/environment-specific tests)
- ❌ **0 failed**

No code changes are required - all tests pass successfully. The PR is ready for review from a testing perspective.

The local `./quality.sh` also passes cleanly: `ok | 1673 passed (2 steps) | 0 failed | 1 ignored`

**Review by nleck — APPROVED** [review]
